lldb/examples/formatter-bytecode/formatter-bytecode.md

```diff
@@ -0,0 +1,165 @@
+
+- `call (Object arg0 ... Selector -> retval)`
+
+Method is one of a predefined set of _Selectors_
+- `(Object @summary -> String)`
+- `(Object @type_summary -> String)`
+
+- `(Object @get_num_children -> UInt)`
+- `(Object UInt @get_child_at_index -> Object)`
+- `(Object String @get_child_index -> UInt)`
+- `(Object @get_type -> Type)`
+- `(Object UInt @get_template_argument_type -> Type)`
+- `(Object @get_value -> Object)`
+- `(Object @get_value_as_unsigned -> UInt)`
+- `(Object @get_value_as_signed -> Int)`
+- `(Object @get_value_as_address -> UInt)`
+- `(Object Type @cast -> Object)`
+
+- `(UInt @read_memory_byte -> UInt)`
+- `(UInt @read_memory_uint32 -> UInt)`
+- `(UInt @read_memory_int32 -> Int)`
+- `(UInt @read_memory_unsigned -> UInt)`
+- `(UInt @read_memory_signed -> Int)`
+- `(UInt @read_memory_address -> UInt)`
+- `(UInt Type @read_memory -> Object)`
+ 
+- `(String arg0 ... fmt -> String)`
+- `(String arg0 ... sprintf -> String)`
+- `(String strlen -> String)`
+
+## Byte Code
+
+Most instructions are just a single byte opcode. The only exceptions are the literals:
+
+- String: Length in bytes encoded as ULEB128, followed length bytes
+- Int: LEB128
+- UInt: ULEB128
+- Selector: ULEB128
+
+### Embedding
+
+Expression programs are embedded into an `.lldbformatters` section (an evolution of the Swift `.lldbsummaries` section) that is a dictionary of type names/regexes and descriptions. It consists of a list of records. Each record starts with the following header:
+
+- version number (ULEB128)
+- remaining size of the record (minus the header) (ULEB128)
+
+Space between two records may be padded with NULL bytes.
+
+In version 1, a record consists of a dictionary key, which is type name or regex.
+
+- length of the key in bytes (ULEB128)
+- the key (UTF-8)
+
+A regex has to start with `^`.
+
+This is followed by one or more dictionary values that immediately follow each other and entirely fill out the record size from the header. Each expression program has the following layout:
+
+- function signature (1 byte)
+- length of the program (ULEB128)
+- the program bytecode
+
+The possible function signatures are:
+- 0: `@summary (Object -> String)`
+- 1: `@init (Object -> Object+)`
+- 2: `@get_num_children (Object+ -> UInt)`
+- 3: `@get_child_index (Object+ String -> UInt)`
+- 4: `@get_child_at_index (Object+ UInt -> Object)`
+- FIXME: potentially also `get_value`? (https://lldb.llvm.org/use/variable.html#synthetic-children)
+
+If not specified, the init function defaults to an empty function that just passes the Object along. Its results may be cached and allow common prep work to be done for an Object that can be reused by subsequent calls to the other methods. This way subsequent calls to `@get_child_at_index` can avoid recomputing shared information, for example.
+
+While it is more efficient to store multiple programs per type key, this is not a requirement. LLDB will merge all entries. If there are conflicts the result is undefined.
```

Comment:
What is the use case for having multiple programs?
```
typename -> [formatter 1, formatter 2]
```
Perhaps?

Are they executed serially to produce one formatted output, or does each one produce one output?

So you could say ok this container can be viewed in this way or this way. Perhaps one for debugging details for the library developer and one for a user who just wants to see their own data.

Reply:
Yes, the synthetic child providers have to define a set of functions, and it would be less efficient to repeat the typename for all of them.

Reply:
> Are they executed serially to produce one formatted output, or does each one produce one output?

No, they serve different purposes based on their signature. (Summary, init, etc...)

Reply:
Right now I see the list of those in the doc.

So if you did want to have 2 different views you'd write a single summary provider that returned both, or arrange for different providers to be built into release and debug builds.

(or more likely just have the debug ones be python scripts given that developers of the library are already knee deep in complexity)

Along the same lines, is it possible to take the providers loaded from the section and assign them to new types? For example, I might subclass a container from a library I use. Would the library have to write their regex carefully to allow that or can I get a handle to the existing provider and attach that to my own type?